File: fpkg_lib/builder.py
from __future__ import annotations

import os
import sys
import shutil
import subprocess
import tempfile
import tomllib
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional
from datetime import datetime, timezone

from .manifest import (
    Manifest, ManifestPackage, ManifestVerification,
    ManifestDependencies, ManifestInstall, ManifestRepository,
    ManifestCompat, PackageSize, PackageFlags,
)
from .package import FpkgWriter


class BuildError(Exception):
    pass


@dataclass
class PkgBuildSource:
    url: str = ""
    sha256: str = ""
    local: str = ""


@dataclass
class PkgBuildRuntime:
    requires: list[str] = field(default_factory=list)
    suggests: list[str] = field(default_factory=list)
    conflicts: list[str] = field(default_factory=list)
    provides: list[str] = field(default_factory=list)


@dataclass
class PkgBuildSection:
    build_depends: list[str] = field(default_factory=list)
    script: str = ""


@dataclass
class PkgBuildInstallSection:
    script: str = ""
    config_files: list[str] = field(default_factory=list)
    desktop_files: list[str] = field(default_factory=list)
    mode: str = "both"


@dataclass
class PkgBuild:
    name: str
    version: str
    release: int = 1
    arch: str = "x86_64"
    license: str = ""
    summary: str = ""
    description: str = ""
    homepage: str = ""
    maintainer: str = ""
    categories: list[str] = field(default_factory=list)
    tags: list[str] = field(default_factory=list)
    source: PkgBuildSource = field(default_factory=PkgBuildSource)
    build: PkgBuildSection = field(default_factory=PkgBuildSection)
    package_install: PkgBuildInstallSection = field(default_factory=PkgBuildInstallSection)
    runtime: PkgBuildRuntime = field(default_factory=PkgBuildRuntime)
    scripts: dict[str, str] = field(default_factory=dict)
    changelog: str = ""

    @classmethod
    def from_file(cls, path: str) -> PkgBuild:
        with open(path, "rb") as f:
            data = tomllib.load(f)
        return cls.from_dict(data)

    @classmethod
    def from_dict(cls, data: dict) -> PkgBuild:
        pkg = data.get("package", {})
        src = data.get("source", {})
        bld = data.get("build", {})
        inst = data.get("package_install", {})
        rt = data.get("runtime", {})
        scr = data.get("scripts", {})

        return cls(
            name=pkg.get("name", ""),
            version=pkg.get("version", ""),
            release=pkg.get("release", 1),
            arch=pkg.get("arch", "x86_64"),
            license=pkg.get("license", ""),
            summary=pkg.get("summary", ""),
            description=pkg.get("description", ""),
            homepage=pkg.get("homepage", ""),
            maintainer=pkg.get("maintainer", ""),
            categories=pkg.get("categories", []),
            tags=pkg.get("tags", []),
            source=PkgBuildSource(
                url=src.get("url", ""),
                sha256=src.get("sha256", ""),
                local=src.get("local", ""),
            ),
            build=PkgBuildSection(
                build_depends=bld.get("build_depends", []),
                script=bld.get("script", ""),
            ),
            package_install=PkgBuildInstallSection(
                script=inst.get("script", ""),
                config_files=inst.get("config_files", []),
                desktop_files=inst.get("desktop_files", []),
                mode=inst.get("mode", "both"),
            ),
            runtime=PkgBuildRuntime(
                requires=rt.get("requires", []),
                suggests=rt.get("suggests", []),
                conflicts=rt.get("conflicts", []),
                provides=rt.get("provides", []),
            ),
            scripts=scr,
            changelog=data.get("changelog", ""),
        )

    def validate(self) -> list[str]:
        errors = []
        if not self.name:
            errors.append("package.name is required")
        if not self.version:
            errors.append("package.version is required")
        if self.arch not in ("x86_64", "aarch64", "riscv64", "any"):
            errors.append("package.arch must be x86_64 | aarch64 | riscv64 | any")
        return errors

    def output_filename(self) -> str:
        return f"{self.name}-{self.version}-{self.release}-{self.arch}.fpkg"


class Builder:
    def __init__(self, pkgbuild: PkgBuild, output_dir: str = ".", verbose: bool = False):
        self.pkgbuild = pkgbuild
        self.output_dir = Path(output_dir)
        self.verbose = verbose

    def _log(self, msg: str):
        print(msg, flush=True)

    def _run_script(self, script: str, env: dict[str, str]):
        if not script.strip():
            return
        result = subprocess.run(
            ["/bin/sh", "-e", "-c", script],
            env={**os.environ, **env},
            capture_output=not self.verbose,
        )
        if result.returncode != 0:
            stderr = result.stderr.decode() if result.stderr else ""
            raise BuildError(f"Script failed (exit {result.returncode}):\n{stderr}")

    def build(self) -> str:
        errors = self.pkgbuild.validate()
        if errors:
            raise BuildError("PKGBUILD validation failed:\n" + "\n".join(f"  - {e}" for e in errors))

        pkg = self.pkgbuild
        output_path = self.output_dir / pkg.output_filename()

        self._log(f"[*] Building {pkg.name} {pkg.version}-{pkg.release} ({pkg.arch})")

        with tempfile.TemporaryDirectory(prefix="fpkg-build-") as tmpdir:
            srcdir = os.path.join(tmpdir, "src")
            destdir = os.path.join(tmpdir, "dest")
            os.makedirs(srcdir, exist_ok=True)
            os.makedirs(destdir, exist_ok=True)

            if pkg.source.local:
                local_src = Path(pkg.source.local)
                if not local_src.exists():
                    raise BuildError(f"Local source not found: {pkg.source.local}")
                if local_src.is_dir():
                    shutil.copytree(str(local_src), srcdir, dirs_exist_ok=True)
                else:
                    shutil.copy2(str(local_src), srcdir)
                self._log(f"[✓] Source: {pkg.source.local}")
            elif pkg.source.url:
                self._log(f"[*] Fetching source: {pkg.source.url}")
                self._fetch_source(pkg.source.url, pkg.source.sha256, srcdir)
                self._log(f"[✓] Source fetched")
            else:
                self._log(f"[!] No source specified, using empty DATA/")

            if pkg.build.script.strip():
                self._log(f"[*] Running build script...")
                self._run_script(pkg.build.script, {
                    "FPM_SRCDIR": srcdir,
                    "FPM_DESTDIR": destdir,
                    "FPM_NAME": pkg.name,
                    "FPM_VERSION": pkg.version,
                    "FPM_ARCH": pkg.arch,
                })
                self._log(f"[✓] Build complete")

            if pkg.package_install.script.strip():
                self._log(f"[*] Running install script...")
                self._run_script(pkg.package_install.script, {
                    "FPM_SRCDIR": srcdir,
                    "FPM_DESTDIR": destdir,
                    "FPM_NAME": pkg.name,
                    "FPM_VERSION": pkg.version,
                    "FPM_ARCH": pkg.arch,
                })
                self._log(f"[✓] Install script complete")

            manifest = self._build_manifest(pkg)
            writer = FpkgWriter(str(output_path))
            writer.set_manifest(manifest)

            data_files_count = 0
            for file_path in Path(destdir).rglob("*"):
                if not file_path.is_file():
                    continue
                rel = file_path.relative_to(destdir)
                writer.add_data_file(str(rel), file_path.read_bytes())
                data_files_count += 1

            for script_name, script_content in pkg.scripts.items():
                writer.set_script(script_name, script_content.encode())

            if pkg.changelog:
                writer.set_changelog(pkg.changelog)

            writer.set_origin("native")

            self._log(f"[*] Packing {data_files_count} file(s)...")
            writer.write()
            self._log(f"[✓] Package created: {output_path}")

        return str(output_path)

    def _fetch_source(self, url: str, expected_sha256: str, destdir: str):
        import urllib.request
        import hashlib

        filename = url.split("/")[-1].split("?")[0] or "source"
        dest_file = os.path.join(destdir, filename)

        urllib.request.urlretrieve(url, dest_file)

        if expected_sha256:
            with open(dest_file, "rb") as f:
                actual = hashlib.sha256(f.read()).hexdigest()
            if actual != expected_sha256:
                raise BuildError(f"SHA256 mismatch: expected {expected_sha256}, got {actual}")

        if filename.endswith((".tar.gz", ".tgz", ".tar.bz2", ".tar.xz", ".tar.zst")):
            import tarfile
            with tarfile.open(dest_file) as tf:
                tf.extractall(destdir)
            os.unlink(dest_file)
        elif filename.endswith(".zip"):
            import zipfile
            with zipfile.ZipFile(dest_file) as zf:
                zf.extractall(destdir)
            os.unlink(dest_file)

    def _build_manifest(self, pkg: PkgBuild) -> Manifest:
        return Manifest(
            package=ManifestPackage(
                name=pkg.name,
                version=pkg.version,
                release=pkg.release,
                arch=pkg.arch,
                license=pkg.license,
                summary=pkg.summary,
                description=pkg.description,
                homepage=pkg.homepage,
                source_url=pkg.source.url or pkg.source.local,
                maintainer=pkg.maintainer,
                build_date=datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
                categories=pkg.categories,
                tags=pkg.tags,
            ),
            verification=ManifestVerification(
                signature_algo="ed25519",
            ),
            dependencies=ManifestDependencies(
                requires=pkg.runtime.requires,
                suggests=pkg.runtime.suggests,
                conflicts=pkg.runtime.conflicts,
                provides=pkg.runtime.provides,
            ),
            install=ManifestInstall(
                mode=pkg.package_install.mode,
                config_files=pkg.package_install.config_files,
                desktop_files=pkg.package_install.desktop_files,
            ),
            repository=ManifestRepository(),
            compat=ManifestCompat(min_fpm_version="0.1.0", fpkg_format="1"),
        )

File: fpkg-build/src/builder.rs
use anyhow::{bail, Context, Result};
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;
use tempfile::TempDir;
use walkdir::WalkDir;

use crate::manifest::{
    Manifest, ManifestCompat, ManifestDependencies, ManifestFlags, ManifestInstall,
    ManifestPackage, ManifestRepository, ManifestSize, ManifestVerification,
};
use crate::package::FpkgWriter;
use crate::pkgbuild::PkgBuild;

pub struct Builder {
    pub output_dir: PathBuf,
    pub verbose: bool,
}

impl Builder {
    pub fn new(output_dir: PathBuf, verbose: bool) -> Self {
        Self { output_dir, verbose }
    }

    pub fn build(&self, pkg: &PkgBuild) -> Result<PathBuf> {
        let output_path = self.output_dir.join(pkg.output_filename());

        println!(
            "[*] Building {} {}-{} ({})",
            pkg.package.name, pkg.package.version, pkg.package.release, pkg.package.arch
        );

        let workdir = TempDir::new().context("Cannot create temp directory")?;
        let srcdir = workdir.path().join("src");
        let destdir = workdir.path().join("dest");
        fs::create_dir_all(&srcdir)?;
        fs::create_dir_all(&destdir)?;

        self.prepare_source(pkg, &srcdir)?;
        self.run_script_phase("build", &pkg.build.script, &srcdir, &destdir, pkg)?;
        self.run_script_phase("install", &pkg.install.script, &srcdir, &destdir, pkg)?;

        let manifest = Self::build_manifest(pkg);
        let mut writer = FpkgWriter::new(output_path.to_string_lossy().to_string(), manifest);

        let mut file_count = 0usize;
        for entry in WalkDir::new(&destdir).into_iter().filter_map(|e| e.ok()) {
            if !entry.file_type().is_file() {
                continue;
            }
            let rel = entry.path().strip_prefix(&destdir)?;
            let content = fs::read(entry.path())?;
            writer.add_data_file(&rel.to_string_lossy(), content);
            file_count += 1;
        }

        Self::attach_scripts(&mut writer, pkg);

        if !pkg.changelog.is_empty() {
            writer.changelog = pkg.changelog.clone();
        }

        println!("[*] Packing {} file(s)...", file_count);
        let compressed = writer.write()?;
        println!(
            "[✓] Package created: {}",
            output_path.display()
        );
        println!("    Compressed: {} bytes", compressed);

        Ok(output_path)
    }

    fn prepare_source(&self, pkg: &PkgBuild, srcdir: &Path) -> Result<()> {
        if !pkg.source.local.is_empty() {
            let local = PathBuf::from(&pkg.source.local);
            if !local.exists() {
                bail!("Local source not found: {}", pkg.source.local);
            }
            if local.is_dir() {
                copy_dir(&local, srcdir)?;
            } else {
                fs::copy(&local, srcdir.join(local.file_name().unwrap()))?;
            }
            println!("[✓] Source: {}", pkg.source.local);
        } else if !pkg.source.url.is_empty() {
            println!("[*] Fetching: {}", pkg.source.url);
            self.fetch_url(&pkg.source.url, &pkg.source.sha256, srcdir)?;
            println!("[✓] Source fetched");
        } else {
            println!("[!] No source — using empty DATA/");
        }
        Ok(())
    }

    fn fetch_url(&self, url: &str, expected_sha256: &str, destdir: &Path) -> Result<()> {
        let filename = url.split('/').last().unwrap_or("source");
        let dest_file = destdir.join(filename);

        let status = Command::new("curl")
            .args(["-fsSL", "-o", &dest_file.to_string_lossy(), url])
            .status()
            .context("curl not found")?;

        if !status.success() {
            bail!("Download failed: {}", url);
        }

        if !expected_sha256.is_empty() {
            let data = fs::read(&dest_file)?;
            let actual = format!("blake3:{}", blake3::hash(&data).to_hex());
            let expected_normalized = if expected_sha256.starts_with("blake3:") {
                expected_sha256.to_string()
            } else {
                format!("sha256:{}", expected_sha256)
            };
            if actual != expected_normalized && format!("sha256:{}", expected_sha256) != expected_normalized {
                bail!("Checksum mismatch for {}", filename);
            }
        }

        let name = filename.to_lowercase();
        if name.ends_with(".tar.gz") || name.ends_with(".tgz")
            || name.ends_with(".tar.bz2") || name.ends_with(".tar.xz")
            || name.ends_with(".tar.zst")
        {
            let status = Command::new("tar")
                .args(["xf", &dest_file.to_string_lossy(), "-C", &destdir.to_string_lossy()])
                .status()
                .context("tar not found")?;
            if !status.success() {
                bail!("Failed to extract archive");
            }
            fs::remove_file(&dest_file)?;
        }

        Ok(())
    }

    fn run_script_phase(
        &self,
        phase: &str,
        script: &str,
        srcdir: &Path,
        destdir: &Path,
        pkg: &PkgBuild,
    ) -> Result<()> {
        if script.trim().is_empty() {
            return Ok(());
        }

        println!("[*] Running {} script...", phase);

        let mut cmd = Command::new("/bin/sh");
        cmd.arg("-e")
            .arg("-c")
            .arg(script)
            .env("FPM_SRCDIR", srcdir)
            .env("FPM_DESTDIR", destdir)
            .env("FPM_NAME", &pkg.package.name)
            .env("FPM_VERSION", &pkg.package.version)
            .env("FPM_ARCH", &pkg.package.arch)
            .env("FPM_MODE", "build");

        if self.verbose {
            let status = cmd.status().context("Failed to run script")?;
            if !status.success() {
                bail!("Script failed (exit {})", status.code().unwrap_or(-1));
            }
        } else {
            let output = cmd.output().context("Failed to run script")?;
            if !output.status.success() {
                let stderr = String::from_utf8_lossy(&output.stderr);
                bail!(
                    "{} script failed (exit {}):\n{}",
                    phase,
                    output.status.code().unwrap_or(-1),
                    stderr
                );
            }
        }

        println!("[✓] {} complete", phase);
        Ok(())
    }

    fn attach_scripts(writer: &mut FpkgWriter, pkg: &PkgBuild) {
        let pairs = [
            ("pre-install.sh", &pkg.scripts.pre_install),
            ("post-install.sh", &pkg.scripts.post_install),
            ("pre-remove.sh", &pkg.scripts.pre_remove),
            ("post-remove.sh", &pkg.scripts.post_remove),
        ];
        for (name, content) in &pairs {
            if !content.is_empty() {
                writer.add_script(name, content.as_bytes().to_vec());
            }
        }
    }

    fn build_manifest(pkg: &PkgBuild) -> Manifest {
        let source_url = if !pkg.source.url.is_empty() {
            pkg.source.url.clone()
        } else {
            pkg.source.local.clone()
        };

        Manifest {
            package: ManifestPackage {
                name: pkg.package.name.clone(),
                version: pkg.package.version.clone(),
                release: pkg.package.release,
                arch: pkg.package.arch.clone(),
                license: pkg.package.license.clone(),
                summary: pkg.package.summary.clone(),
                description: pkg.package.description.clone(),
                homepage: pkg.package.homepage.clone(),
                source_url,
                maintainer: pkg.package.maintainer.clone(),
                build_date: Manifest::build_date(),
                categories: pkg.package.categories.clone(),
                tags: pkg.package.tags.clone(),
                size: ManifestSize { installed: 0, compressed: 0, delta_base: String::new() },
                flags: ManifestFlags {
                    system_config: false,
                    has_services: false,
                    selinux_aware: false,
                    has_suid: false,
                },
            },
            verification: ManifestVerification {
                manifest_hash: String::new(),
                content_tree: String::new(),
                signature_algo: "ed25519".into(),
            },
            dependencies: ManifestDependencies {
                requires: pkg.runtime.requires.clone(),
                suggests: pkg.runtime.suggests.clone(),
                conflicts: pkg.runtime.conflicts.clone(),
                provides: pkg.runtime.provides.clone(),
                before: Vec::new(),
                after: Vec::new(),
            },
            install: ManifestInstall {
                mode: pkg.install.mode.clone(),
                config_files: pkg.install.config_files.clone(),
                desktop_files: pkg.install.desktop_files.clone(),
            },
            repository: ManifestRepository {
                origin_repo: String::new(),
                origin_url: String::new(),
                fetched_at: String::new(),
            },
            compat: ManifestCompat {
                min_fpm_version: "0.1.0".into(),
                fpkg_format: "1".into(),
            },
        }
    }
}

fn copy_dir(src: &Path, dst: &Path) -> Result<()> {
    for entry in WalkDir::new(src).into_iter().filter_map(|e| e.ok()) {
        let rel = entry.path().strip_prefix(src)?;
        let target = dst.join(rel);
        if entry.file_type().is_dir() {
            fs::create_dir_all(&target)?;
        } else {
            if let Some(parent) = target.parent() {
                fs::create_dir_all(parent)?;
            }
            fs::copy(entry.path(), &target)?;
        }
    }
    Ok(())
}
